### PR TITLE
LX-1396 umount failed (part deux)

### PR DESCRIPTION
**Pull request by grwilson**
This change provides a coarse-grain file lock to the `zfs` and `zpool` commands to prevent concurrent updates to the `/etc/dfs/sharetab` file.  The lock is grabbed whenever we to call `zfs set sharenfs` or when we `zpool destroy` or `zpool export` the pool. This is meant as a hack to allow our testing to complete successfully as it does not protect against concurrent updates when commands use `-o sharenfs` property to update the nfs settings. We rely on the fact that our stack currently only sets the `sharenfs` property by calling `zfs set`.

### CODE REVIEW

File: cmd/zfs/zfs_main.c [hunk 1]
```diff
@@ -3904,9 +3904,27 @@ zfs_do_set(int argc, char **argv)
 		}
 	}
 
+	/*
+	 * Lock the sharetab to ensure concurrent writers don't update
+	 * the file simultaneously.
```
Comment (ahrens) [inline cmd/zfs/zfs_main.c]: [nit] prefix this with `If we are setting the sharenfs property, lock...`

File: cmd/zpool/zpool_main.c [hunk 2]
```diff
@@ -1603,8 +1606,12 @@ zpool_export_one(zpool_handle_t *zhp, void *data)
 {
 	export_cbdata_t *cb = data;
 
-	if (zpool_disable_datasets(zhp, cb->force) != 0)
+	verify(sharetab_lock() == 0);
+	if (zpool_disable_datasets(zhp, cb->force) != 0) {
+		verify(sharetab_unlock() == 0);
```
Comment (ahrens) [inline cmd/zpool/zpool_main.c]: I realize that this whole thing is a hack, but would it be useful to lock when doing `zpool import` as well?  (AFAIK the product doesn't use either export or import, but maybe relevant to testing?)

File: lib/libshare/libshare.c [hunk 3]
```diff
@@ -503,9 +502,6 @@ sa_fini(sa_handle_t handle)
 
 	update_sharetab(impl_handle);
 
-	if (impl_handle->zfs_libhandle != NULL)
-		libzfs_fini(impl_handle->zfs_libhandle);
```
Comment (ahrens) [inline lib/libshare/libshare.c]: can you explain this change?

Comment (grwilson) [reply lib/libshare/libshare.c]: I hit an issue here when I was trying to add asserts that the sharetab_lock was held. I will revert this change but I found this code path very sketchy:

Comment (grwilson) [reply lib/libshare/libshare.c]: BTW, I gave up adding asserts because the call flow is not sane.

**Comment by grwilson** [comment]
It's possible that this problem would show up in production. Effectively it requires the stack to share out two or more timeflows concurrently.

**Comment by prakashsurya** [comment]
I'm a little curious why we don't need to hold the lock until `libzfs_fini` is called from `main`, which is one place where `update_sharetab` will be called? i.e. I thought it might be possible for the cache to get initialized when a share is processed, then this cache to get written out when `libzfs_fini` is called, so we'd need to hold the lock up until the point `libzfs_fini` (i.e. `zfs_uninit_libshare` -> `sa_fini`). Am I mistaken?

**Comment by grwilson** [comment]
@prakashsurya and I chatted about closing the window down completely and his suggestion was to make the locking variable global so that we can drop it after we call `libzfs_fini` in `main`.

**Comment by grwilson** [comment]
One other options is to just avoid the unlock code and allow that to happen automatically when the process terminates. Would you guys like to see an explicit unlock or just rely on process termination for this hack?

**Comment by ahrens** [comment]
It seems cleaner to explicitly drop the lock.  But assuming that it's all in zfs_main.c / zpool_main.c, I'd be OK with dropping it via process termination, especially if that would make it simple enough that we could actually cover all the cases (e.g. `-o sharenfs=`).

**Comment by prakashsurya** [comment]
I’m good with either approach.

**Comment by grwilson** [comment]
I will look at covering all of the `-o sharenfs`. I think that should be possible.

**Comment by grwilson** [comment]
The latest changes will cover all of the `zfs -o sharenfs` and `zfs set` cases.

**Pull request merged**